nautobot/ipam/api/views.py

```diff
@@ -405,36 +411,86 @@ def available_vlans(self, request, pk=None):
         By default, the number of VIDs returned will be equivalent to PAGINATE_COUNT.
         An arbitrary limit (up to MAX_PAGE_SIZE, if set) may be passed, however results will not be paginated.
         """
-        obj = get_object_or_404(self.queryset, pk=pk)
+        vlan_group = get_object_or_404(self.queryset, pk=pk)
 
-        try:
-            limit = int(request.query_params.get("limit", get_settings_or_config("PAGINATE_COUNT")))
-        except ValueError:
-            limit = get_settings_or_config("PAGINATE_COUNT")
+        if request.method == "POST":
+            with cache.lock(
+                "nautobot.ipam.api.views.available_vlans", blocking_timeout=5, timeout=settings.REDIS_LOCK_TIMEOUT
+            ):
+                # Normalize to a list of objects
+                requested_vlans = request.data if isinstance(request.data, list) else [request.data]
+
+                # Determine if the requested number of VLANs is available
+                available_vids = vlan_group.available_vids
+                if len(available_vids) < len(requested_vlans):
+                    return Response(
+                        {
+                            "detail": (
+                                f"An insufficient number of VLANs are available within the VLANGroup {vlan_group} "
+                                f"({len(requested_vlans)} requested, {len(available_vids)} available)"
+                            )
+                        },
+                        status=status.HTTP_204_NO_CONTENT,
+                    )
+
+                # Assign VLAN IDs from the list of VLANGroup's available VLAN IDs.
+                _available_vids = iter(available_vids)
+
+                for requested_vlan in requested_vlans:
+                    if "vid" not in requested_vlan:
+                        requested_vlan["vid"] = f"{next(_available_vids)}"
+
+                    requested_vlan["vlan_group"] = vlan_group.pk
```

Comment:
May need special case handling for custom field data here (see #6131), similar to what we do for the available-prefixes and available-ips endpoints:

```py
                    # The serializer usage above has mapped "custom_fields" dict to "_custom_field_data".
                    # We need to convert it back to "custom_fields" as we're going to deserialize it a second time below
                    requested_vlan["custom_fields"] = requested_vlan.pop("_custom_field_data", {})
```

Or is this not necessary because we're not passing `request.data` through a first-round deserialization above (see my other comment)?

Reply:
It is not needed as we return `serializer.data` and not `serializer_validated_data`.
We do not run the first deserialization as in case of `VLAN` it is not possible to fully validate it without a VID.